### ISSUE TITLE
[2.1.0] BaseCSVWidget.value_from_datadict — AttributeError: 'list' object has no attribute 'split'

### ISSUE THREAD

**Issue by moseb** [carltongibson/django-filter#1103]
I ran into …
```
Traceback (most recent call last):
  [..]
  File "/usr/local/lib/python3.7/site-packages/django/forms/forms.py", line 393, in _clean_fields
    value = field.widget.value_from_datadict(self.data, self.files, self.add_prefix(name))
  File "/usr/local/lib/python3.7/site-packages/django_filters/widgets.py", line 201, in value_from_datadict
    return value.split(',')
AttributeError: 'list' object has no attribute 'split'
```
and found that the code at …
https://github.com/carltongibson/django-filter/blob/1f47e36b614724a8735e0457fa511dcaf5448481/django_filters/widgets.py#L195-L202
… is not robust with regard to `super().value_from_datadict(data, files, name)` returning a list. I need to debug this more but maybe you already know what's going on here?

**Comment by moseb** [comment]
Upstream ticket: https://github.com/carltongibson/django-filter/issues/1103

**Comment by rpkilby** [comment]
Hi @moseb. Can you paste the FilterSet code that you're using? The last few lines of the trace back are sufficient to determine why you'd hit this exception.

**Comment by moseb** [comment]
Full traceback:
```

**Comment by moseb** [comment]
It's from browsing a listing where django-rest-framework generates the code

**Comment by moseb** [comment]
> Hi @moseb. Can you paste the FilterSet code that you're using?

To be honest, I haven't found any in that area of the code, yet. It might be all by DRF. Not sure yet.
I'm aware my bug report is not ideal but it's the best I have right now. Sorry!

**Comment by rpkilby** [comment]
Your API view should have either a `filterset_class` or `filterset_fields`, or if you're using an older version of django-filter, `filter_class` and `filter_fields`.

**Comment by moseb** [comment]
I found use of `filter_class` — so I guess you just helped me find that the [migration to django-filter 2.0](https://django-filter.readthedocs.io/en/master/guide/migration.html#migrating-to-2-0) was not properly done to its end. Cool!
Is there something stopping django-filter from detecting pre-2.x leftovers and warning about them, exceptionally? Does it warn but I missed it? Would code warning about these be a welcome addition for a pull request?

**Comment by rpkilby** [comment]
v2.0 removed a lot of the existing deprecation warnings, so you might try running your test suite against v1.0 and then v1.1 of django-filter.

**Comment by carltongibson** [comment]
Assuming this will resolve itself when you update. Come back if not.

**Comment by moseb** [comment]
It turns out it's not an django-filter-2.0-migration issue. The issue surfaces with use of `rest_framework_filters.backends.RestFrameworkFilterBackend` of https://github.com/philipn/django-rest-framework-filters version 1.0.0.dev0 rather than plain `django_filters.rest_framework.DjangoFilterBackend` of django-filter, it seems. I'm not exactly sure yet why.

**Comment by moseb** [comment]
There is reproducible demo of the issue now at https://github.com/moseb/django-filter-issue-1103-demo. Please consider re-opening this issue. Thank you!

**Comment by rpkilby** [comment]
Ah, this is related to `ModelMultipleChoiceFilter` not being compatible with the CSV mixin, as well as the `in` lookup. Two separate but related issues here.

You shouldn't need an `in` lookup for m2m fields, since the multiple choice filter already provides similar behavior.

**Comment by moseb** [comment]
Hi!

> Ah, this is related to `ModelMultipleChoiceFilter` not being compatible with the CSV mixin, as well as the `in` lookup. Two separate but related issues here.

Is there existing tickets on GitHub for these issues?

It sounds like making `BaseCSVWidget.value_from_datadict` more robust as mentioned above alone will not fix the issue? (I don't fully understand the issue yet.)

> You shouldn't need an `in` lookup for m2m fields, since the multiple choice filter already provides similar behavior.

`in` allows checking against multiple values over `exact`. If I cannot use `ìn`, how do I check for multiple values?

PS: Can we re-open this ticket?

**Comment by rpkilby** [comment]
> It sounds like making `BaseCSVWidget.value_from_datadict` more robust as mentioned above alone will not fix the issue? (I don't fully understand the issue yet.)

Possibly. If the value is already a list, it should just go ahead and return that list. That said, I don't think it makes sense to mix the CSV-behavior with the `SelectMultiple` widget.

> `in` allows checking against multiple values over `exact`. If I cannot use `ìn`, how do I check for multiple values?

The `ModelMultipleChoiceFilter` constructs an OR query from Q objects. So, if you have a query string like `/api/mymodel?m2m=a&m2m=b`, you would end up with a filter call like

```python
MyModel.objects.filter(Q(m2m='a') | Q(m2m='b'))
```

With the 2.1 release, the `lookup_expr` is now applied to each Q object, so you'd end up with 

```python
MyModel.objects.filter(Q(m2m__in='a') | Q(m2m__in='b'))
```

The above is invalid and would break, however it makes sense for `contains` and other lookups that are intended to work with single values.

So, in short, all you need to do here is use the `exact` lookup for your m2m field.

----

I think the todo here is:
- Make `BaseCSVWidget` compatible with `SelectMultiple` by returning the value if it's already a list. Or if it shouldn't be compatible, we should at least provide a useful error on init.
- Fix `Meta.fields` for m2m. Generating an `in` lookup for `ModelMultipleChoiceFilter` is nonsensical. We should also test what happens for other lookups like `isnull`, which expects a boolean.

**Comment by carltongibson** [comment]
Yeah, this still doesn't look like a **bug** but how it all works is under documented I guess...

**Comment by marcosox** [comment]
I just had this error and its explanation is buried in this thread, so to recap and save a long search for future readers:

`ManyToMany` fields declared in viewsets' `filterset_fields` **must not** contain `in`

A warning at runtime would be very appreciated.

**Comment by carltongibson** [comment]
@marcosox Good idea. Happy to look at a PR adding that!
